### PR TITLE
DRILL-4539: Add support for Null Equality Joins

### PR DESCRIPTION
**Pull request by vkorukanti**
@amansinha100 @jacques-n Could you please review the patches?

### CODE REVIEW

**Comment by jacques-n** [comment]
@amansinha100, can you take a look?

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillRelOptUtil.java [hunk 1]
```diff
@@ -169,4 +176,223 @@ private static boolean containIdentity(List<? extends RexNode> exps,
     }
     return true;
   }
+
+  /**
+   * Copied from {@link RelOptUtil#splitJoinCondition(RelNode, RelNode, RexNode, List, List)}. Modified to rewrite
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillRelOptUtil.java]: Is there a way to make this not copied? @julianhyde, any ideas for how to refactor common stuff out?

Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillRelOptUtil.java]: Agree that we ideally should leverage the Calcite code..especially since this method is pretty heavily used and modified periodically so keeping Drill's version of this method in sync will be difficult.

File: exec/java-exec/src/main/codegen/templates/ComparisonFunctions.java [hunk 2]
```diff
@@ -215,6 +192,36 @@ public void eval() {
     }
   }
 
+  <#-- IS_DISTINCT_FROM function -->
+  @FunctionTemplate(names = {"is_distinct_from", "is distinct from" },
```
Comment (amansinha100) [inline exec/java-exec/src/main/codegen/templates/ComparisonFunctions.java]: @vkorukanti, I want to clarify...if the query only had a join condition with IS_NOT_DISTINCT_FROM, I would think it should work just with your convertlet changes, since both HashJoin and MergeJoin handle this type of join condition.  Is the reason you had to implement the full comparator codegen to handle more general types of comparisons ?  e.g in the SELECT list if I say  'SELECT  a IS NOT DISTINCT FROM b'  ?    Suppose we had a convertlet that only preserved the IS (NOT) DISTINCT FROM join condition, and defaulted to the Calcite rewrite using the CASE expression, then we would not have to implement the full comparator.

Comment (vkorukanti) [inline exec/java-exec/src/main/codegen/templates/ComparisonFunctions.java]: I am not sure if there is way to differentiate between the function in join condition vs. function in project expr. I don't see any context info in DrillConvertletTable.get() method call. Also the generated code in rewritten case is too much. For following query:
```SELECT INT_col is not distinct from BIGINT_col as col, int_distinct_result FROM cp.`functions/distinct_from.json```

Without rewrite: https://gist.github.com/vkorukanti/e981058f985ed24e6c4ef6b47d670e0f
With rewrite: https://gist.github.com/vkorukanti/d80aa2ba40c65c9215c38ed18b20a685

Sizes may differ after scalar replacement is done, but it is still too much code to simple ```is not distinct from``` function.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillRelOptUtil.java (hunk 1 above)
Comment (vkorukanti) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillRelOptUtil.java]: I will followup with a JIRA on Calcite project to see if we can push this change to Calcite. 

The function ```RelOptUtil#splitJoinCondition``` in the current form itself seems to have a problem/limitation. Currently it just returns the left and right join key indices, but doesn't return whether the condition is ```EQUAL``` or ```IS NOT DISTINCT FROM``` (it adds the key pair if they have either of these function in comparison).

File: exec/java-exec/src/main/codegen/templates/ComparisonFunctions.java (hunk 2 above)
Comment (vkorukanti) [inline exec/java-exec/src/main/codegen/templates/ComparisonFunctions.java]: If adding new functions is a concern, I can make the ```RelOptUtil#splitJoinCondition``` to identify rewritten ```IS NOT DISTINCT FROM``` functions also.

Comment (amansinha100) [inline exec/java-exec/src/main/codegen/templates/ComparisonFunctions.java]: I am not opposed to having a native implementation of IS [NOT] DISTINCT FROM...clearly the generated code is more compact; however adding this new functions means we would need proper functional test coverage for various data types.  Any thoughts regarding that ?

Comment (vkorukanti) [inline exec/java-exec/src/main/codegen/templates/ComparisonFunctions.java]: I added tests for each category of template code path (primitive type, decimal type and interval type) in TestIsDistinctFromFunctions.java

File: exec/java-exec/src/test/java/org/apache/drill/TestJoinNullable.java [hunk 3]
```diff
@@ -407,11 +342,94 @@ public void testMergeLOJNullableBothInputsOrderedDescNullsLastVsAscNullsLast() t
...
+        .baselineValues("A", "L_A_2", "R_A_1", "A")
+        .baselineValues(null, "L_null_1", "R_null_2", null)
+        .baselineValues(null, "L_null_2", "R_null_2", null)
+        .baselineValues(null, "L_null_1", "R_null_3", null)
+        .baselineValues(null, "L_null_2", "R_null_3", null)
+        .go();
   }
 
+  @Test
+  public void withNullEqualAdditionFilter() throws Exception {
```
Comment (amansinha100) [inline exec/java-exec/src/test/java/org/apache/drill/TestJoinNullable.java]: Could you also do similar test with the join condition in the WHERE clause instead of ON clause ?  i.e something like:  SELECT * FROM t1, t2 WHERE t1.a = t2.a OR (t1.a is null and t2.a is null)
For such cases, Calcite filter pushdown into join needs to be applied first.

Comment (vkorukanti) [inline exec/java-exec/src/test/java/org/apache/drill/TestJoinNullable.java]: Sure. I will update the patch with new tests.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillRelOptUtil.java [hunk 4]
```diff
@@ -169,4 +176,223 @@ private static boolean containIdentity(List<? extends RexNode> exps,
...
+
+    return RexUtil.composeConjunction(
+        left.getCluster().getRexBuilder(), nonEquiList, false);
+  }
+
+  /**
+   * Copied from {@link RelOptUtil#splitJoinCondition(int, RexNode, List, List, List)}. Modified to rewrite the null
+   * equal join condition using IS NOT DISTINCT FROM operator.
+   */
+  private static void splitJoinCondition(
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillRelOptUtil.java]: Can you confirm if this rewrite does *not* do the conversion if the join condition happens to involve columns coming from not just 2 tables but from 3  tables ? e.g if the user accidentally gives: 

SELECT * FROM t1, t2, t3 WHERE t1.a = t2.a  OR (t1.b is null and t3.b is null)

Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillRelOptUtil.java]: Rest looks good to me.  +1.

**Comment by vkorukanti** [comment]
Added few more tests:
```SELECT * FROM t1, t2, t3 WHERE t1.a = t2.a OR (t1.b is null and t3.b is null)``` to test it throws CannotPlanException
```SELECT * FROM t1, t2 WHERE t1.a = t2.a OR (t1.a is null and t2.a is null)``` to test null-equality filter is pushed into join condition. Had to make some changes to ```DrillFilterJoinRules``` to handle this case.

Also few more tests around 3 table join that involve null-equality condition in ```WHERE``` clause and ```JOIN ON```

4 tests from extended regression suite failed, because now we generate a join condition that involves ```=`` and ```IS NOT DISTINCT FROM```. 

```SELECT * FROM cp.`jsonInput/nullableOrdered1.json` t1 JOIN cp.`jsonInput/nullableOrdered2.json` t2 ON t1.key = t2.key WHERE t1.data IS NOT DISTINCT FROM t2.data```

It looks like HashJoin/MergeJoin operators are not handling mixed comparator in join condition. Looking at the ```ChainedHashTable``` and ```MergeJoinBatch``` code briefly, it seems like we should be able to generate code for handling mixed comparators case. @amansinha100: Let me know if I am not correct.

**Comment by amansinha100** [comment]
Right, the HashJoin and MergeJoin currently can handle one type of comparison, not mixed...this is why when we have a join condition such as  t1.a1 = t2.a2  AND t2.b1 < t2.b2  then Drill planner currently will split it up into equality join and the non-equality conditions.  The equality is handled by the join operator while a new Filter node is added after the join to process the non-equality.   I had created a JIRA to be able to handle both types of comparisons in the join operator:  see https://issues.apache.org/jira/browse/DRILL-3803.    Do you need this support for this particular null equality join issue ?  I would have thought that your rewrite rule would produce the IS NOT DISTINCT FROM condition, but not a mixed condition.

**Comment by vkorukanti** [comment]
I haven't tried this, but wondering whether this is sufficient to handle mix of EQUALS and IS NOT DISTINCT FROM cases.  Maintain a list (one for each join condition) instead of single "areNullsEquals" flag in MergeJoinBatch [1] and ChainedHashTable [2] and generate the code accordingly for each condition.

[1] https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/MergeJoinBatch.java#L463

[2] https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/ChainedHashTable.java#L251

**Comment by amansinha100** [comment]
I think additional support may be needed; the hash table only stores the equality join keys, so the generated code in ChainedHashTable is targeted to the those columns.  For the additional columns you will need to keep track of them (I haven't thought through the details yet.)

**Comment by vkorukanti** [comment]
It seems like mixed condition of EQUALS and IS NOT DISTINCT FROM in the same join condition can be supported few changes. Here are the WIP changes [1]. Need to test more and clean it up.

[1] https://github.com/vkorukanti/drill/commit/0c8a4a2cef926492853e9c8ae856b82615be3c72

**Comment by vkorukanti** [comment]
Updated PR to handle mixed EQUALS and IS NOT DISTINCT FROM in same join condition. @amansinha100 Could you please review the latest commit in the PR?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashTableConfig.java [hunk 5]
```diff
@@ -22,25 +22,36 @@
...
+                         @JsonProperty("keyExprsProbe") NamedExpression[] keyExprsProbe,
+                         @JsonProperty("comparators") JoinComparator[] comparators) {
     this.initialCapacity = initialCapacity;
     this.loadFactor = loadFactor;
     this.keyExprsBuild = keyExprsBuild;
     this.keyExprsProbe = keyExprsProbe;
+    if (comparators != null) {
+      this.comparators = comparators;
+    } else {
+      // default is nulls are equal
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashTableConfig.java]: It might seem confusing at first (it was to me) to consider the nulls-are-equal as the default case since that's not true for joins; however it is true for a group-by operation.  Could you have the caller pass in explicit comparator such that the HashTableConfig does not have to decide ?  The HashAggregate could instantiate the hash table with an IS_NOT_DISTINCT_FROM comparator.  What do you think ?  (one minor point is that the 'JoinComparator' is now being used for both Joins and Aggregates...we could call it Comparator or something generic).

**Comment by amansinha100** [comment]
I had one comment on the latest commit but overall LGTM.  +1

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashTableConfig.java (hunk 5 above)
Comment (vkorukanti) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashTableConfig.java]: Agree, it is clean to remove passing null and explicitly pass a value to comparator. Updated patch. Also renamed join.JoinComparator to common.Comparator.
